### ISSUE TITLE
When Sandstorm shell discovers "session headers" have changed, make a new session

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#643]
(Question: Will this result in a new `X-Sandstorm-Session-Id`? If so, what can an app rely on as a persistent user session identifier?)

**Comment by kentonv** [comment]
The app should never be relying on `X-Sandstorm-Session-Id` for semantic sessions, because there are lots of reasons it could change: the Cap'n Proto connection could be dropped and re-established, the grain could die and restart, etc.

Outside of Sandstorm, there is actually no reasonable way to implement a session identifier! There are _user_ identifiers, but they are shared across the browser, not tab-specific. To associate requests from one tab, the app has to manually generate a session identifier and make sure to include it with every request, e.g. in the URLs or in headers. It doesn't work very well.

Inside Sandstorm, the app can use the hostname, or cookies, or localStorage, as a semantic session identifier. So we actually provide a feature here that "legacy" web doesn't. :)

**Comment by paulproteus** [comment]
I agree with all of that. However, outside of Sandstorm, there is a stable
hostname and therefore basically-stable session cookies shared across the
browser, and that's the feature I want help replicating in Sandstorm.

I think the answer we discussed once was to make the per-tab session
subdomains longer-lasting, which works for me.

**Comment by kentonv** [comment]
When do you want a "browser-wide" session but not a "user-wide" session? I would argue never. :)

**Comment by mitar** [comment]
> When do you want a "browser-wide" session but not a "user-wide" session? I would argue never. :)

When user is anonymous? ;-)
